File: go/shield/fingerprint.go
package shield

import (
	"crypto/md5"
	"fmt"
	"os"
	"os/exec"
	"runtime"
	"strings"
)

// FingerprintMode defines the fingerprint collection mode.
type FingerprintMode int

const (
	// FingerprintNone - No fingerprinting (backward compatible)
	FingerprintNone FingerprintMode = iota
	// FingerprintMotherboard - Motherboard serial only
	FingerprintMotherboard
	// FingerprintCPU - CPU identifier only
	FingerprintCPU
	// FingerprintCombined - Motherboard + CPU (recommended)
	FingerprintCombined
)

// CollectFingerprint collects hardware fingerprint based on mode.
//
// Platform Support:
//   - Windows: wmic commands
//   - Linux: /sys/class/dmi/id, dmidecode, /proc/cpuinfo
//   - macOS: system_profiler, sysctl
//
// Returns:
//   - Fingerprint string (MD5 hex), or empty for FingerprintNone
//   - Error if hardware identifiers cannot be collected
func CollectFingerprint(mode FingerprintMode) (string, error) {
	if mode == FingerprintNone {
		return "", nil
	}

	if mode == FingerprintMotherboard {
		return getMotherboardSerial()
	}

	if mode == FingerprintCPU {
		return getCpuID()
	}

	if mode == FingerprintCombined {
		var components []string

		if mb, err := getMotherboardSerial(); err == nil {
			components = append(components, mb)
		}

		if cpu, err := getCpuID(); err == nil {
			components = append(components, cpu)
		}

		if len(components) == 0 {
			return "", fmt.Errorf("hardware fingerprint unavailable")
		}

		// Create MD5 hash of combined components
		combined := strings.Join(components, "-")
		return fmt.Sprintf("%x", md5.Sum([]byte(combined))), nil
	}

	return "", fmt.Errorf("unknown fingerprint mode: %d", mode)
}

// getMotherboardSerial gets motherboard serial number (platform-specific).
func getMotherboardSerial() (string, error) {
	switch runtime.GOOS {
	case "windows":
		out, err := exec.Command("wmic", "baseboard", "get", "serialnumber", "/value").Output()
		if err != nil {
			return "", err
		}

		for _, line := range strings.Split(string(out), "\n") {
			if strings.HasPrefix(line, "SerialNumber=") {
				serial := strings.TrimSpace(strings.TrimPrefix(line, "SerialNumber="))
				if serial != "" && serial != "To be filled by O.E.M." {
					return serial, nil
				}
			}
		}

	case "linux":
		// Try DMI sysfs first (no elevated privileges needed)
		data, err := os.ReadFile("/sys/class/dmi/id/board_serial")
		if err == nil {
			serial := strings.TrimSpace(string(data))
			if serial != "" && serial != "To be filled by O.E.M." {
				return serial, nil
			}
		}

		// Fallback to dmidecode
		out, err := exec.Command("dmidecode", "-s", "baseboard-serial-number").Output()
		if err == nil {
			serial := strings.TrimSpace(string(out))
			if serial != "" && serial != "To be filled by O.E.M." {
				return serial, nil
			}
		}

	case "darwin":
		out, err := exec.Command("system_profiler", "SPHardwareDataType").Output()
		if err != nil {
			return "", err
		}

		for _, line := range strings.Split(string(out), "\n") {
			if strings.Contains(line, "Serial Number") {
				parts := strings.Split(line, ":")
				if len(parts) >= 2 {
					return strings.TrimSpace(parts[1]), nil
				}
			}
		}
	}

	return "", fmt.Errorf("motherboard serial number unavailable")
}

// getCpuID gets CPU identifier (platform-specific).
func getCpuID() (string, error) {
	switch runtime.GOOS {
	case "windows":
		out, err := exec.Command("wmic", "cpu", "get", "ProcessorId", "/value").Output()
		if err != nil {
			return "", err
		}

		for _, line := range strings.Split(string(out), "\n") {
			if strings.HasPrefix(line, "ProcessorId=") {
				cpuID := strings.TrimSpace(strings.TrimPrefix(line, "ProcessorId="))
				if cpuID != "" {
					return cpuID, nil
				}
			}
		}

	case "linux":
		data, err := os.ReadFile("/proc/cpuinfo")
		if err == nil {
			for _, line := range strings.Split(string(data), "\n") {
				if strings.HasPrefix(line, "processor") && strings.Contains(line, "0") {
					// Use first processor as identifier (hashed)
					return fmt.Sprintf("%x", md5.Sum([]byte(line))), nil
				}
			}
		}

	case "darwin":
		out, err := exec.Command("sysctl", "-n", "machdep.cpu.brand_string").Output()
		if err == nil {
			cpuInfo := strings.TrimSpace(string(out))
			if cpuInfo != "" {
				return fmt.Sprintf("%x", md5.Sum([]byte(cpuInfo))), nil
			}
		}
	}

	return "", fmt.Errorf("CPU identifier unavailable")
}

// NewWithFingerprint creates a Shield instance with hardware fingerprinting.
//
// Derives keys from password + hardware identifier, binding encryption to
// the physical device.
//
// Example:
//
//	shield, err := shield.NewWithFingerprint("password", "github.com", shield.FingerprintCombined)
//	if err != nil {
//	    log.Fatal(err)
//	}
//	encrypted, _ := shield.Encrypt([]byte("secret data"))
func NewWithFingerprint(password, service string, mode FingerprintMode) (*Shield, error) {
	fingerprint, err := CollectFingerprint(mode)
	if err != nil {
		return nil, err
	}

	combinedPassword := password
	if fingerprint != "" {
		combinedPassword = fmt.Sprintf("%s:%s", password, fingerprint)
	}

	return New(combinedPassword, service, nil), nil
}

File: python/shield/fingerprint.py
"""Hardware fingerprinting for device-bound encryption.

Collects platform-specific hardware identifiers to create device-bound keys.
Adapted from SaaSClient-SideLicensingSystem with enhanced cross-platform support.
"""

import hashlib
import platform
import subprocess
from enum import Enum
from typing import Optional


class FingerprintMode(Enum):
    """Fingerprint collection mode."""

    NONE = "none"  # No hardware fingerprinting (backward compatible)
    MOTHERBOARD = "motherboard"  # Use motherboard serial number only
    CPU = "cpu"  # Use CPU identifier only
    COMBINED = "combined"  # Use combined motherboard + CPU (recommended)


class FingerprintError(Exception):
    """Hardware fingerprint unavailable (VM or restricted access)."""

    pass


def collect_fingerprint(mode: FingerprintMode) -> str:
    """Collect hardware fingerprint based on mode.

    Args:
        mode: Fingerprint collection mode

    Returns:
        Hardware fingerprint string (empty for FingerprintMode.NONE)

    Raises:
        FingerprintError: If hardware identifiers cannot be collected

    Platform Support:
        - Windows: wmic commands (baseboard, CPU)
        - Linux: /sys/class/dmi/id, dmidecode, /proc/cpuinfo
        - macOS: system_profiler SPHardwareDataType
    """
    if mode == FingerprintMode.NONE:
        return ""

    if mode == FingerprintMode.MOTHERBOARD:
        return _get_motherboard_serial()

    if mode == FingerprintMode.CPU:
        return _get_cpu_id()

    if mode == FingerprintMode.COMBINED:
        components = []

        try:
            components.append(_get_motherboard_serial())
        except FingerprintError:
            pass

        try:
            components.append(_get_cpu_id())
        except FingerprintError:
            pass

        if not components:
            raise FingerprintError("No hardware identifiers available")

        # Create MD5 hash of combined components (matches SaaSClient and Rust)
        combined = "-".join(components)
        return hashlib.md5(combined.encode()).hexdigest()

    raise ValueError(f"Unknown fingerprint mode: {mode}")


def _get_motherboard_serial() -> str:
    """Get motherboard serial number (platform-specific)."""
    system = platform.system()

    if system == "Windows":
        try:
            result = subprocess.run(
                ["wmic", "baseboard", "get", "serialnumber", "/value"],
                capture_output=True,
                text=True,
                check=False,
            )
            for line in result.stdout.splitlines():
                if line.startswith("SerialNumber="):
                    serial = line.replace("SerialNumber=", "").strip()
                    if serial and serial != "To be filled by O.E.M.":
                        return serial
        except Exception:
            pass

    elif system == "Linux":
        # Try DMI sysfs first (no elevated privileges needed)
        try:
            with open("/sys/class/dmi/id/board_serial", "r") as f:
                serial = f.read().strip()
                if serial and serial != "To be filled by O.E.M.":
                    return serial
        except Exception:
            pass

        # Fallback to dmidecode (may require sudo)
        try:
            result = subprocess.run(
                ["dmidecode", "-s", "baseboard-serial-number"],
                capture_output=True,
                text=True,
                check=False,
            )
            serial = result.stdout.strip()
            if serial and serial != "To be filled by O.E.M.":
                return serial
        except Exception:
            pass

    elif system == "Darwin":  # macOS
        try:
            result = subprocess.run(
                ["system_profiler", "SPHardwareDataType"],
                capture_output=True,
                text=True,
                check=False,
            )
            for line in result.stdout.splitlines():
                if "Serial Number" in line:
                    parts = line.split(":")
                    if len(parts) >= 2:
                        return parts[1].strip()
        except Exception:
            pass

    raise FingerprintError("Motherboard serial number unavailable")


def _get_cpu_id() -> str:
    """Get CPU identifier (platform-specific)."""
    system = platform.system()

    if system == "Windows":
        try:
            result = subprocess.run(
                ["wmic", "cpu", "get", "ProcessorId", "/value"],
                capture_output=True,
                text=True,
                check=False,
            )
            for line in result.stdout.splitlines():
                if line.startswith("ProcessorId="):
                    cpu_id = line.replace("ProcessorId=", "").strip()
                    if cpu_id:
                        return cpu_id
        except Exception:
            pass

    elif system == "Linux":
        try:
            with open("/proc/cpuinfo", "r") as f:
                for line in f:
                    if line.startswith("processor") and "0" in line:
                        # Use first processor as identifier (hashed)
                        return hashlib.md5(line.encode()).hexdigest()
        except Exception:
            pass

    elif system == "Darwin":  # macOS
        try:
            result = subprocess.run(
                ["sysctl", "-n", "machdep.cpu.brand_string"],
                capture_output=True,
                text=True,
                check=False,
            )
            cpu_info = result.stdout.strip()
            if cpu_info:
                return hashlib.md5(cpu_info.encode()).hexdigest()
        except Exception:
            pass

    raise FingerprintError("CPU identifier unavailable")
